File: regression/linear_regression.R
# This script creates and evaluates a linear regression model in R for the 'iris' data.
# This iteration is only to get the model up and running, so there is no feature engineering and parameter tuning.

# LOAD LIBRARIES ----------------------------------------------------------
# This section loads relevant libraries for this script.
library(readr)
library(data.table)
library(caret)

# IMPORT DATA -------------------------------------------------------------
# This section imports data.
# The data files are located in the 'R' branch file, NOT the 'classification' or 'regression' folders.
train = data.table(read_csv('train.csv'))
test = data.table(read_csv('test.csv'))

# CREATE MODEL ------------------------------------------------------------
# This section creates a linear regression model.

  # Set seed for reproducibility.
  set.seed(555)

  # Create linear regression model and include 5 fold cross validation.
  linear_reg = train(Sepal.Length ~ Sepal.Width + Petal.Length + Petal.Width, 
                     train, 
                     method = 'lm', 
                     trControl = trainControl(method = 'cv', 
                                              number = 5))
  
  # View model summary.
  summary(linear_reg)
  
  # View variable importance plot.
  plot(varImp(linear_reg))
  
  # Generate training and test set predictions
  train$linear_reg_pred = predict(linear_reg, train)
  test$linear_reg_pred = predict(linear_reg, 
                                 test)

# EVALUATE MODEL ----------------------------------------------------------
# This section evaluates the model by computing training, cross validation, and test set root mean square error (RMSE).
# Note that I do not include degrees of freedom in the training and test set calculations below.
# I'm not sure how 'caret' computes RMSE in the cross validation calculation.
  
  # Compute training set RMSE
  sqrt(sum((train$Sepal.Length - predict(linear_reg)) ^ 2) / nrow(train))
  
  # Compute cross validation set RMSE.
  mean(linear_reg$resample$RMSE)
  
  # Compute test set RMSE.
  sqrt(sum((test$Sepal.Length - test$linear_reg_pred) ^ 2) / nrow(test))

File: regression/linear_regression.py
# This script creates and evaluates a linear regression model in Python for the 'iris' data.
# This iteration is only to get the model up and running, so there is no feature engineering and parameter tuning.

################
# IMPORT MODULES
################
# This section imports modules needed for this script.
import os
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error

#############
# IMPORT DATA
#############
# This section imports data.
# The data files are located in the 'Python' branch file, NOT the 'classification' or 'regression' folders.
# The commands below assume the working directory is set to the 'classification' folder.
os.chdir('..')
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

##############
# CREATE MODEL
##############
# This section creates a linear regression model.

# Create and fit linear regression model.
linear_reg = LinearRegression().fit(train[['Sepal.Width', 'Petal.Length', 'Petal.Width']],
                                    train['Sepal.Length'])

# Conduct 5 fold cross validation.
cv = cross_val_score(linear_reg,
                     train[['Sepal.Width', 'Petal.Length', 'Petal.Width']],
                     train['Sepal.Length'],
                     cv = 5,
                     scoring = 'neg_mean_squared_error')

# View model coefficients.
linear_reg.coef_

# View intercept.
linear_reg.intercept_

# Generate training and test set predictions.
train['linear_reg_pred'] = linear_reg.predict(train[['Sepal.Width', 'Petal.Length', 'Petal.Width']])
test['linear_reg_pred'] = linear_reg.predict(test[['Sepal.Width', 'Petal.Length', 'Petal.Width']])

################
# EVALUATE MODEL
################
# This section computes the mean squared error of the model on the training, cross validation, and test sets.

# Compute mean squared error on training set.
mean_squared_error(train['Sepal.Length'],
                   train['linear_reg_pred'])

# Compute mean squared error on cross validation set.
cv.mean() * -1

# Compute mean squared error on test set.
mean_squared_error(test['Sepal.Length'],
                   test['linear_reg_pred'])
